File: solutions/scala/triangle/1/src/main/scala/Triangle.scala
case class Triangle(sides: List[Number]) {
  lazy val equilateral: Boolean = isValidTriangle &&
                                  hasAllEqualSides
  lazy val isosceles  : Boolean = isValidTriangle &&
                                  hasAtLeastTwoEqualSides
  lazy val scalene    : Boolean = isValidTriangle &&
                                  hasNoEqualSides
  private lazy val eachValueExceedsSumOfAllOthers: Boolean = {
    val sortedSides = sides.map(_.floatValue()).sorted
    sortedSides.head + sortedSides(1) > sortedSides(2)
  }
  private lazy val hasNoEqualSides        : Boolean = sides.distinct.length == 3
  private lazy val hasAllEqualSides       : Boolean = sides.distinct.length == 1
  private lazy val hasAtLeastTwoEqualSides: Boolean = sides.distinct.length <= 2
  private lazy val isValidTriangle: Boolean = !sides.contains(0) &&
                                              eachValueExceedsSumOfAllOthers
}

object Triangle {
  def apply(side1: Number, side2: Number, side3: Number): Triangle = new Triangle(List(side1, side2, side3))
}

File: solutions/python/triangle/1/triangle.py
def is_valid_triangle(sides):
    first_side, second_side, third_side = sides
    has_nonpositive_side = (
            first_side <= 0
            or second_side <= 0
            or third_side <= 0
    )
    if has_nonpositive_side:
        return False

    return (
            first_side + second_side >= third_side
            and second_side + third_side >= first_side
            and first_side + third_side >= second_side
    )

def equilateral(sides):
    if not is_valid_triangle(sides):
        return False

    first_side, second_side, third_side = sides
    return first_side == second_side == third_side


def isosceles(sides):
    if not is_valid_triangle(sides):
        return False

    first_side, second_side, third_side = sides
    return (
        first_side == second_side
        or first_side == third_side
        or second_side == third_side
    )


def scalene(sides):
    if not is_valid_triangle(sides):
        return False

    first_side, second_side, third_side = sides
    return (
        first_side != second_side
        and first_side != third_side
        and second_side != third_side
    )